functorch/experimental/_cond.py

```diff
@@ -56,12 +56,35 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
 
     flat_true_outs, _ = pytree.tree_flatten(true_outs)
     flat_false_outs, _ = pytree.tree_flatten(false_outs)
-    assert(len(flat_true_outs) == len(flat_false_outs))
+    if len(flat_true_outs) != len(flat_false_outs):
+        raise UserError(
+            UserErrorType.COND_OP_RESTRICTION,
+            "Expect true branch and false branch return with same length but got\n"
+            "  {true_fn} returns {true_outs}\n"
+            "  {false_fn} returns {false_outs}\n".format(
+                true_fn=true_fn.__name__,
+                true_outs=flat_true_outs,
+                false_fn=false_fn.__name__,
+                false_outs=flat_false_outs
+            )
+        )
 
     for i in range(0, len(flat_true_outs)):
         true_out = flat_true_outs[i]
         false_out = flat_false_outs[i]
-        assert true_out.meta['tensor_meta'] == false_out.meta['tensor_meta']
+        if true_out.meta['tensor_meta'] != false_out.meta['tensor_meta']:
+            raise UserError(
+                UserErrorType.COND_OP_RESTRICTION,
+                "Expect each tensor returned from true branch and false branch "
```

Comment:
Expect[ed] each tensor...[to have] [CUT exact] same metadata but [CUT got]

Reply:
Is metadata user exposed concept? What would user even do from seeing metadata? Maybe this should just be an assertion.

Reply:
Hmm, you're right that metadata is not really something the user needs to understand. But each of the individual components of metadata are, right? Like, shape, dtype, etc. are things user should understand, and it's OK to complain when they don't match across branches.